nuxeo-features/nuxeo-automation/nuxeo-automation-core/src/main/java/org/nuxeo/ecm/automation/core/operations/document/AddPermission.java

```diff
@@ -81,88 +95,111 @@
-        ACE ace = ACE.builder(username, permission)
-                     .creator(creator)
-                     .begin(begin)
-                     .end(end)
-                     .contextData(contextData)
-                     .build();
         boolean permissionChanged = false;
         if (blockInheritance) {
             permissionChanged = acp.blockInheritance(aclName, creator);
         }
-        permissionChanged = acp.addACE(aclName, ace) || permissionChanged;
+
+        for (String username : users) {
+            ACE ace = ACE.builder(username, permission)
+                         .creator(creator)
+                         .begin(begin)
+                         .end(end)
+                         .contextData(contextData)
+                         .build();
+            permissionChanged = acp.addACE(aclName, ace) || permissionChanged;
+        }
+
         if (permissionChanged) {
             doc.setACP(acp, true);
         }
     }
 
-    protected void checkUserExistence(String username) {
-        UserManager userManager = Framework.getService(UserManager.class);
-        if (userManager.getUserModel(username) == null && userManager.getGroupModel(username) == null) {
-            String errorMsg = "User or group name '" + username + "' does not exist. Please provide a valid name.";
+    protected void validateParameters() {
+        if (user == null && (users == null || users.isEmpty()) && email == null) {
+            throw new IllegalParameterException("'users' or 'email' parameters must be set");
+        } else if (email != null && end == null) {
+            throw new IllegalParameterException("'end' parameter must be set when adding a permission for an 'email'");
+        }
+
+        ConfigurationService configService = Framework.getService(ConfigurationService.class);
+        if (configService.isBooleanPropertyFalse(ALLOW_VIRTUAL_USER)) {
+            UserManager userManager = Framework.getService(UserManager.class);
+            if (user != null) {
+                checkUserAndGroupExistence(user, userManager);
+            }
+            if (users != null && !users.isEmpty()) {
+                for (String username : users) {
+                    checkUserAndGroupExistence(username, userManager);
+                }
+            }
+        }
+    }
+
+    protected void checkUserAndGroupExistence(String userOrGroupName, UserManager userManager) {
+        if (userManager.getUserModel(userOrGroupName) == null && userManager.getGroupModel(userOrGroupName) == null) {
+            String errorMsg = String.format("User or group name '%s' does not exist. Please provide a valid name.",
+                    userOrGroupName);
             throw new IllegalParameterException(errorMsg);
         }
     }
 
+    /**
+     * Method to help deprecating {@link #user} parameter.
+     */
+    protected void ensureUserListIsUsed() {
+        users = users == null ? new ArrayList<>() : new ArrayList<>(users);
```

Comment:
Why do you need the copy if it's not null?

Reply:
If the parameter passed was something like "Arrays.asList(...)", the size is fixed and we cannot add values to that list.

Reply:
Hm you're right of course. I thought these params were always reconstructed from scratch by the Automation framework but there's no reason for this to be the case.